### PR TITLE
rework: raven running raven

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
Fixes the mechanics for raven-runs-raven, and adds a simple test for the same.

The complex test is currently failing due to problems with the inner RAVEN; I'm still working on fixing that.

However, the test added proves the RRR mechanics do work as expected, and will enable it to be used and benchmarked on the dataobject-rework branch.

### CODE REVIEW

File: doc/tests/attenuate.tex [hunk 1]
```diff
@@ -96,3 +96,41 @@ \subsubsection{numeric values}
 \begin{equation}
 \text{var}{u(Y)} = 0.063779804051749989
 \end{equation}
+
+\subsection{Changing lower, upper bounds}
+A parametric study can be made by changing the lower and upper bounds of the
```
Comment (alfoa) [inline doc/tests/attenuate.tex]: ```of the```?

File: framework/Models/Code.py [hunk 2]
```diff
@@ -544,13 +544,49 @@ def evaluateSample(self, myInput, samplerType, kwargs):
...
+          rlz.update( new )
+          ## combine ProbabilityWeights # TODO FIXME these are a rough attempt at getting it right!
+          rlz['ProbabilityWeight'] = np.atleast_1d(rlz.get('ProbabilityWeight',1.0) * kwargs.get('ProbabilityWeight',1.0))
+          rlz['PointProbability'] = np.atleast_1d(rlz.get('PointProbability',1.0) * kwargs.get('PointProbability',1.0))
+          rlz['prefix'] = np.atleast_1d(kwargs['prefix']+'_'+str(n))
+          ## add the rest of the metadata # TODO slow
+          for var,val in kwargs.items():
+            if var not in rlz.keys():
+              rlz[var] = np.atleast_1d(val)
+          self._replaceVariablesNamesWithAliasSystem(rlz,'inout',True) # TODO should this only be "input" instead of "inout"?
```
Comment (alfoa) [inline framework/Models/Code.py]: inout is correct

File: framework/Models/Code.py [hunk 3]
```diff
@@ -544,13 +544,49 @@ def evaluateSample(self, myInput, samplerType, kwargs):
...
+          exportDict['realizations'].append(rlz)
+
+          #### OLD METHOD, kept for reference ####
+          #self._replaceVariablesNamesWithAliasSystem(dataObj.getVars('input'), 'input', True)
+          # TODO unstructured inputs not a part of ensemble anymore? FIXME
+          #self._replaceVariablesNamesWithAliasSystem(dataObj.getParametersValues('unstructuredinputs',nodeId='RecontructEnding'), 'input', True)
+          #self._replaceVariablesNamesWithAliasSystem(dataObj.getVars('output'), 'output', True)
+        #returnDict = finalCodeOutputFile
```
Comment (alfoa) [inline framework/Models/Code.py]: indentation

File: framework/DataObjects/XDataSet.py [hunk 4]
```diff
@@ -526,6 +524,41 @@ def remove(self,realization=None,variable=None):
...
+    """
+    # determine where the old variable was
+    isInput = old in self._inputs
+    isOutput = old in self._outputs
+    isMeta = old in self._metavars
+    isIndex = old in self.indexes
+    # make the changes to the variable listings
+    if isInput:
+      self._inputs = list(a if (a != old) else new for a in self._inputs)
+    if isOutput:
```
Comment (alfoa) [inline framework/DataObjects/XDataSet.py]: even if in terms of performance should be the same, I would prefer to have ```elif``` statements for more clarity of the flow logic.

File: doc/tests/attenuate.tex (hunk 1 above)
Comment (PaulTalbot-INL) [reply doc/tests/attenuate.tex]: "material opacities", fixed.

File: framework/Models/Code.py (hunk 2 above)
Comment (PaulTalbot-INL) [reply framework/Models/Code.py]: comment removed

File: framework/Models/Code.py (hunk 3 above)
Comment (PaulTalbot-INL) [reply framework/Models/Code.py]: removed.

File: framework/DataObjects/XDataSet.py (hunk 4 above)
Comment (alfoa) [reply framework/DataObjects/XDataSet.py]: ok. no problem to me

File: tests/framework/CodeInterfaceTests/raven_running_raven_internal_models/test_rom_trainer.xml [hunk 5]
```diff
@@ -160,7 +192,7 @@
     </PointSet>
     <PointSet name="outputMontecarloRom">
       <Input>DeltaTimeScramToAux,DG1recoveryTime</Input>
-      <Output>CladTempThreshold</Output>
+      <!--Output>CladTempThreshold</Output-->
```
Comment (wangcj05) [inline tests/framework/CodeInterfaceTests/raven_running_raven_internal_models/test_rom_trainer.xml]: rm?

File: framework/Models/Code.py [hunk 6]
```diff
@@ -544,13 +544,41 @@ def evaluateSample(self, myInput, samplerType, kwargs):
...
+        ## set up the return container
+        exportDict = {'RAVEN_isBatch':True,'realizations':[]}
+        ## set up each realization
+        for n in range(numRlz):
+          rlz = {}
+          ## collect the results from INNER, both point set and history set
+          for dataObj in finalCodeOutputFile.values():
+            # TODO FIXME check for overwriting data.  For now just replace data if it's duplicate!
+            new = dict((var,np.atleast_1d(val)) for var,val in dataObj.realization(index=n,unpackXArray=True).items())
+            rlz.update( new )
```
Comment (wangcj05) [inline framework/Models/Code.py]: How do you handle the output data from both PointSet and HistorySet with the same name? Will the code crash?

File: tests/framework/CodeInterfaceTests/raven_running_raven_internal_models/test_rom_trainer.xml (hunk 5 above)
Comment (PaulTalbot-INL) [reply tests/framework/CodeInterfaceTests/raven_running_raven_internal_models/test_rom_trainer.xml]: This is part of the unfinished fixing, as mentioned in the description.  It is ongoing work.

File: framework/Models/Code.py (hunk 6 above)
Comment (PaulTalbot-INL) [reply framework/Models/Code.py]:
```python
# TODO FIXME check for overwriting data.  For now just replace data if it's duplicate!
```
That's exactly what this explains.  At some point we will need to decide the action to take; for now, we do one step better than the old implementation and update values in the realization dictionary, which may cause overwriting.

Comment (wangcj05) [reply framework/Models/Code.py]: Based on the discussion, this part will be improved in the future.

File: tests/framework/CodeInterfaceTests/raven_running_raven_internal_models/test_rom_trainer.xml (hunk 5 above)
Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/raven_running_raven_internal_models/test_rom_trainer.xml]: ok

File: framework/Samplers/AdaptiveSobol.py [hunk 7]
```diff
@@ -400,6 +400,7 @@ def localGenerateInput(self,model,oldInput):
         self.inputInfo['ProbabilityWeight']*=self.inputInfo['ProbabilityWeight-'+dist]
     self.inputInfo['PointProbability'] = reduce(mul,self.inputInfo['SampledVarsPb'].values())
     # reassign SampledVarsPb to fully correlated variables
+<<<<<<< HEAD
```
Comment (alfoa) [inline framework/Samplers/AdaptiveSobol.py]: leftover of merging

Comment (PaulTalbot-INL) [reply framework/Samplers/AdaptiveSobol.py]: yes, actively working on this one right now.

**Comment by PaulTalbot-INL** [comment]
Finally, after some mergefixes and rewrites:
```
**** SUMMARY ****
*********************************
**** Unmatched tests are:

framework/CodeInterfaceTests.RAVENrunningRAVEN_ROM
*********************************
**** THE FOLLOWING TESTS CHANGED OUTCOMES ****
**** Test framework/CodeInterfaceTests.RAVENrunningRAVEN | has changed outcome from <FAILED> to <OK> !!!!
*****************
**** THE FOLLOWING TESTS HAVE BEEN ADDED IN 2ND FILE ****
**** Test framework/CodeInterfaceTests.RAVENrunningRAVEN_ROM | has been ADDED  ! It is present in 2nd file BUT NOT in 1st file !!!!
*****************
Number of success first file:
428
Number of success second file:
429
NUMBER OF SUCCESSES INCREASED: +1
**** NUMBER OF TESTS IN THE TWO FILES ARE DIFFERENT: ****
**** first file :533 /= second file: 534****
*********************************
```

**Comment by wangcj05** [comment]
Good to go

**Pull request merged**